### ISSUE TITLE
convenient posterior simulation

### ISSUE THREAD

**Issue by wds15** [stan-dev/rstan#390]
#### Summary:
Add a function to rstan which enables seamless simulation from the posterior draws.

#### Description:
Its a bit cumbersome to evaluate a function per draw of the posterior at the moment. I am proposing to include a function which takes away lots of the usual boiler plate code needed usually.

#### Reproducible Steps:
The the attached files. The basic syntax to simulate over a posterior with some function calc_mean will then be:

```
## the posterior simulate function is clever enough to infer from the
## R function definition of calc_mean what it needs and it will
## automatically grab alpha and beta from the posterior and x from the
## global NS
post_sim1 <- posterior_simulate(post, calc_mean)
```

#### Current Output:
If applicable, any relevant output from RStan.
[posterior_simulate_example.zip](https://github.com/stan-dev/rstan/files/763089/posterior_simulate_example.zip)

#### Expected Output:
If applicable, the output you expected from RStan.

#### RStan Version:
2.14

#### R Version:
3.2.3

#### Operating System:
RHEL 6.7

**Comment by wds15** [comment]
Bump.

If this is of interest I can go ahead and implement it on a branch.

Just some feedback if this should land in rstan would be great - before I start.

I use this function very often, so I thought this is useful to others as well.

**Comment by bgoodri** [comment]
Yeah, Andrew requested something similar. Let me think about the best way
to implement something like this.

**Comment by avehtari** [comment]
I'm worried about the part grabbing something out of global NS. This has caused a lot of problems elsewhere, and it would be safer to require explicit data argument (as will be the case in rstanarm).

**Comment by wds15** [comment]
@bgoodri Great. Note that I did on purpose not parallelize the simulation as one cannot use exposed Stan functions inside forked processes unfortunately. However, the exposed functions work much faster than a "Fixed_param" run (which is similar) due to much less overhead implied by Fixed_param. Let me know if you need a hand. 

@avehtari I found grabbing from the global NS very convenient, but I see your point. In my function you can optionally pass in an environment where things should be taken from (or at least that could be solution; I forgot if I did that or wanted to do it...).
